Code/src/main/java/nl/utwente/group10/ui/components/blocks/Block.java

```diff
@@ -31,124 +44,258 @@
+    protected IntegerProperty connectionState;
+    
+    /** Property for the VisualState. */
+    protected IntegerProperty visualState;
+    
+    /** Property for the expression freshness. */
+    private BooleanProperty exprIsDirty;
 
     /**
+     * Constructs a new Block.
      * @param pane
      *            The pane this block belongs to.
      */
     public Block(CustomUIPane pane) {
         parentPane = pane;
-
-        parentPane.selectedBlockProperty()
-                .addListener(
-                        event -> {
-                            if (parentPane.getSelectedBlock().isPresent()
-                                    && this.equals(parentPane
-                                            .getSelectedBlock().get())) {
-                                this.getStyleClass().add("selected");
-                            } else {
-                                this.getStyleClass().removeAll("selected");
-                            }
-                        });
-
+        connectionState = new SimpleIntegerProperty(ConnectionCreationManager.getConnectionState());
+        visualState = new SimpleIntegerProperty(ConnectionCreationManager.getConnectionState());
+        exprIsDirty = new SimpleBooleanProperty(true);
+        
+        // Add listeners to the states.
+        // Invalidate listeners give the Block a change to react on the state
+        // change before it is cascaded.
+        // Cascade listeners make sure state changes are correctly propagated.
+        connectionState.addListener(a -> invalidateConnectionState());
+        connectionState.addListener(a -> setExprIsDirty(true));
+        connectionState.addListener(this::cascadeConnectionState);
+        visualState.addListener(a -> invalidateVisualState());
+        visualState.addListener(this::cascadeVisualState);
+        
+        // Visually react on selection.
+        parentPane.selectedBlockProperty().addListener(event -> {
+            if (parentPane.getSelectedBlock().isPresent() && this.equals(parentPane.getSelectedBlock().get())) {
+                this.getStyleClass().add("selected");
+            } else {
+                this.getStyleClass().removeAll("selected");
+            }
+        });
+        // Add selection trigger.
         this.addEventHandler(MouseEvent.MOUSE_RELEASED, this::handleMouseEvent);
 
+        // Create the CircleMenu after fully creating the Block.
         Platform.runLater(this::createCircleMenu);
     }
 
+    /**
+     * Creates the CircleMenu for this Block.
+     */
     protected void createCircleMenu() {
         circleMenu = new CircleMenu(this);
     }
 
     /**
-     * Sets this block as the selected block. When this block has already been
-     * selected it spawns a {@link CircleMenu} instead.
+     * Sets this block as the selected block.
+     * A right click also opens the CircleMenu.
      */
     private void handleMouseEvent(MouseEvent t) {
+        parentPane.setSelectedBlock(this);
```

Comment:
That is not entirely true, this method can be used for any MouseEvent. Leaving it this way